src/lib.rs

```diff
@@ -1,6 +1,98 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+//! Manages system resources that can be allocated to VMs and their devices.
+
 #![deny(missing_docs)]
-//! Dummy crate needs high-level documentation.
-/// Dummy public function needs documentation.
-pub fn it_works() {
-    assert!(true);
+
+mod id_allocator;
+
+pub use crate::id_allocator::IdAllocator;
+use std::fmt;
+use std::result;
+
+/// Error type for IdAllocator usage.
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    /// All ids from the range specified are exhausted.
+    Overflow,
+    /// An id that is not part of the specified range was requested to be released.
+    OutOfRange(u32),
+    /// An id that was already released was requested to be released.
+    AlreadyReleased(u32),
+    /// An id  that was never allocated was requested to be released.
+    NeverAllocated(u32),
+    /// There are no more IDs available in the manage range
+    ResourceExhausted,
+    /// The range to manage is invalid.
+    InvalidRange(u32, u32),
+}
+
+impl std::error::Error for Error {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        None
+    }
+}
+
+/// Wrapper over std::result::Result
+pub type Result = result::Result<u32, Error>;
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use Error::*;
+
+        match *self {
+            Overflow => write!(f, "Id counter overflowed."),
+            OutOfRange(id) => write!(f, "Specified id: {} is not in the range.", id),
+            ResourceExhausted => {
+                write!(f, "There are no more available ids in the specified range.")
+            }
+            AlreadyReleased(id) => write!(f, "Specified id: {} is already released.", id),
+            NeverAllocated(id) => write!(
+                f,
+                "Specified id: {} was never allocated, can't release it",
+                id
+            ),
+            InvalidRange(begin, end) => {
+                write!(f, "The range specified: {}-{} is not valid.", begin, end)
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::Error as AllocError;
+    use std::error::Error;
+
+    #[test]
+    fn test_error_messages() {
```

Comment:
We avoid these kind of coverage tests in rust-vmm because we find them to break easily and don't bring a lot of value otherwise. Would you mind removing them if you don't feel strongly about it?